src/hdf5_time_series_storage.jl

```diff
@@ -667,72 +656,48 @@ end
-    attributes::Dict{String, Any},
+    type::Type{<:CONSTANT},
     rows::UnitRange{Int},
-    columns::UnitRange{Int},
 )
-    data = SortedDict{Dates.DateTime, Vector{PiecewiseLinearPointData}}()
-    initial_timestamp = attributes["start_time"]
-    interval = attributes["interval"]
-    start_time = initial_timestamp + interval * (columns.start - 1)
-    if length(columns) == 1
-        data[start_time] = retransform_hdf_array(dataset[rows, columns.start, :, :], type)
-    else
-        data_read = retransform_hdf_array(dataset[rows, columns, :, :], type)
-        for (i, it) in
-            enumerate(range(start_time; length = length(columns), step = interval))
-            data[it] = @view data_read[1:length(rows), i]
-        end
-    end
+    data = retransform_hdf_array(dataset[rows], type)
     return data
 end
 
 function get_hdf_array(
     dataset,
-    type::Union{Type{<:CONSTANT}, Type{LinearFunctionData}},
+    T::Union{
+        Type{LinearFunctionData},
+        Type{QuadraticFunctionData},
+        Type{PiecewiseLinearData},
+    },
     rows::UnitRange{Int},
 )
-    data = retransform_hdf_array(dataset[rows], type)
-    return data
-end
-
-function get_hdf_array(dataset, type::Type{QuadraticFunctionData}, rows::UnitRange{Int})
-    data = retransform_hdf_array(dataset[rows, :, :], type)
-    return data
-end
-
-function get_hdf_array(dataset, type::Type{PiecewiseLinearPointData}, rows::UnitRange{Int})
-    data = retransform_hdf_array(dataset[rows, :, :, :], type)
+    colons = repeat([:], ndims(dataset) - 1)
+    data = retransform_hdf_array(dataset[rows, colons...], T)
     return data
 end
 
 function retransform_hdf_array(data::Array, ::Type{<:CONSTANT})
     return data
 end
 
-function retransform_hdf_array(data::Array, ::Type{LinearFunctionData})
-    return LinearFunctionData.(data)
-end
-
-function retransform_hdf_array(data::Array, T::Type{QuadraticFunctionData})
-    row, column, tuple_length = get_data_dims(data, T)
-    if isnothing(column)
-        t_data = Array{Tuple{Float64, Float64}}(undef, row)
-        for r in 1:row
-            t_data[r] = tuple(data[r, 1:tuple_length]...)
-        end
-    else
-        t_data = Array{Tuple{Float64, Float64}}(undef, row, column)
-        for r in 1:row, c in 1:column
-            t_data[r, c] = tuple(data[r, c, 1:tuple_length]...)
-        end
-    end
-    return _quadratic_from_tuple.(t_data)
+function retransform_hdf_array(
+    data::Array,
+    T::Union{Type{LinearFunctionData}, Type{QuadraticFunctionData}},
+)
+    length_req = fieldcount(get_raw_data_type(T))
+    (size(data)[end] == length_req) || throw(
+        ArgumentError(
+            "Last dimension of data must have length $length_req, got size $(size(data))",
+        ),
+    )
+    dims_to_keep = Tuple(1:(ndims(data) - 1))
+    # Pop off the last dimension and call the constructor on that data
+    return map(x -> T(x...), eachslice(data; dims = dims_to_keep))
```

Comment:
Test the performance of this. This is a function that would be called a lot and I am not sure if this is better than the previous approach allocating the array and filling it.